index.d.ts

```diff
@@ -0,0 +1,170 @@
+    value: IFuture<L, VR>
+  ): IFuture<L, R>
+
+  export function attempt<L, R>(f: () => L): IFuture<L, R>
+
+  export function bimap<A, B, AA, BB>(
+    left: (a: A) => AA,
+    right: (a: B) => BB
+  ): IFuture<AA, any>
+
+  export function both<L, A, B>(
+    left: IFuture<L, A>,
+    right: IFuture<L, B>
+  ): IFuture<L, [A, B]>
+
+  export function cache<L, R>(source: IFuture<L, R>): IFuture<L, R>
+
+  export function chain<L, R, RR>(
+    mapper: (value: R) => IFuture<L, RR>,
+    source: IFuture<L, R>
+  ): IFuture<L, RR>
+
+  export function done<L, R>(
+    fn: (cb: (err: L | null, value: R) => void) => void
+  ): IFuture<L, R>
+
+  export function encase2<L, R, A, B>(
+    fn: (a: A, b: B) => L,
+    a: A,
+    b: B
+  ): IFuture<L, R>
+
+  export function encase3<L, R, A, B, C>(
+    fn: (a: A, b: B, c: C) => L,
+    a: A,
+    b: B,
+    c: C
+  ): IFuture<L, R>
+
+  export function encase<L, R>(fn: () => R): {(): IFuture<L, R>}
+  export function encase<L, R, V>(fn: (v: V) => R, t: V): IFuture<L, R>
+  export function encase<L, R, V>(fn: (v: V) => R): {(t: V): IFuture<L, R>}
+
+  // TODO: Need Help
+  export function encaseN2<L, R>(): IFuture<L, R>
+
+  // TODO: Need Help
+  export function encaseN3<L, R>(): IFuture<L, R>
+
+  // TODO: Need Help
+  export function encaseN<L, R>(): IFuture<L, R>
+
+  // TODO: Need Help
+  export function encaseP2<L, R>(): IFuture<L, R>
+
+  // TODO: Need Help
+  export function encaseP3<L, R>(): IFuture<L, R>
+
+  export function encaseP<L, R>(fn: () => Promise<R>): IFuture<L, R>
+
+  // TODO: Need Help
+  export function extractLeft<L, R>(): IFuture<L, R>
+
+  // TODO: Need Help
+  export function extractRight<L, R>(): IFuture<L, R>
+
+  export function fold<L, R, M>(a: (x: L) => M, b: (x: R) => M): IFuture<any, M>
+
+  // TODO: Need Help
+  export function fork<L, R>(): IFuture<L, R>
+
+  // TODO: Need Help
+  export function go<L, R>(): IFuture<L, R>
+
+  // TODO: Need Help
+  export function hook<L, R>(): IFuture<L, R>
+
+  export function isFuture<L, R>(value: any): boolean
+
+  // TODO: Need Help
+  export function isNever<L, R>(): boolean
+
+  // TODO: Need Help
+  export function lastly<L, R>(): IFuture<L, R>
+
+  export function map<L, R, M>(
+    mapper: (value: L) => M,
+    source: IFuture<L, R>
+  ): IFuture<M, R>
+  export function map<L, R, M>(
+    mapper: (value: L) => M
+  ): ({(source: IFuture<L, R>): IFuture<M, R>})
+
+  // TODO: Need Help
+  export function mapRej<L, R>(): IFuture<L, R>
+
+  // TODO: Need Help
+  export function never<L, R>(): IFuture<L, R>
+
+  // TODO: Need Help
+  export function node<L, R>(): IFuture<L, R>
+
+  export function of<L, R>(value: L): IFuture<L, R>
+
+  export function or<L, R>(a: IFuture<L, R>, b: IFuture<L, R>): IFuture<L, R>
+
+  // TODO: Need Help
+  export function par<L, R>(): IFuture<L, R>
+
+  export function parallel<L, R>(
+    concurrency: number,
+    futures: Array<IFuture<any, any>>
+  ): IFuture<L, Array<any>>
```

Comment:
I *really* prefer to force developers to use consistent types. Again, the fact that it's *possible* in Fluture is simply because it has a huge performance impact to impose this restriction at run-time.

> Like making multiple HTTP requests all of them returning different json responses

I would like for Fluture to move developers to normalize the different responses before passing the futures into `parallel`. Futures that resolve to different types of values should have never made it into the same list.

The same goes for the rejection branch. Once you end up with a Future of a mixed type in your rejection branch, you're going to have a bad time handling that in `fork`. My preference is still for changing this signature to:

```ts
parallel<L, R>(
  concurrency: number,
  futures: Array<IFuture<L, R>>
): IFuture<L, Array<R>>
```